azurerm/data_source_image.go

```diff
@@ -92,37 +108,103 @@ func dataSourceArmImageRead(d *schema.ResourceData, meta interface{}) error {
 	client := meta.(*ArmClient).imageClient
 
 	resGroup := d.Get("resource_group_name").(string)
+
 	name := d.Get("name").(string)
+	nameRegex, nameRegexOk := d.GetOk("name_regex")
+
+	if name == "" && !nameRegexOk {
+		return fmt.Errorf("[ERROR] either name or name_regex is required")
+	}
+
+	var img compute.Image
 
-	resp, err := client.Get(resGroup, name, "")
-	if err != nil {
-		if utils.ResponseWasNotFound(resp.Response) {
+	if !nameRegexOk {
+		var err error
+		if img, err = client.Get(resGroup, name, ""); err != nil {
+			if utils.ResponseWasNotFound(img.Response) {
+				d.SetId("")
+				return nil
+			}
+			return fmt.Errorf("[ERROR] Error making Read request on Azure Image %q (resource group %q): %+v", name, resGroup, err)
+		}
+	} else {
+		r := regexp.MustCompile(nameRegex.(string))
+
+		list := []compute.Image{}
+		resp, err := client.ListByResourceGroup(resGroup)
+		if err != nil {
+			if utils.ResponseWasNotFound(resp.Response) {
+				d.SetId("")
+				return nil
+			}
+			return fmt.Errorf("[ERROR] Error getting list of images (resource group %q): %+v", resGroup, err)
+		}
+		for _, ri := range *resp.Value {
+			if r.Match(([]byte)(*ri.Name)) {
+				list = append(list, ri)
+			}
+		}
+		for resp.NextLink != nil && *resp.NextLink != "" {
+			resp, err = client.ListByResourceGroupNextResults(resp)
+			if err != nil {
+				return fmt.Errorf("[ERROR] Unable to query next results for image list (resource group %q): %+v", resGroup, err)
+			}
+			for _, ri := range *resp.Value {
+				if r.Match(([]byte)(*ri.Name)) {
+					list = append(list, ri)
+				}
+			}
+		}
+
+		if len(list) < 1 {
 			d.SetId("")
 			return nil
 		}
-		return fmt.Errorf("[ERROR] Error making Read request on Azure Image %q (resource group %q): %+v", name, resGroup, err)
+
+		if len(list) > 1 {
+			desc := d.Get("name_descending").(bool)
+			log.Printf("[DEBUG] arm_image - multiple results found and `name_descending` is set to: %t", desc)
```

Comment:
for the AWS provider we fail in this case unless `most_recent` is specified - is it worth mirroring that behaviour?